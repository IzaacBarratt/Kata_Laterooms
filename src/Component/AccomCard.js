import React from 'react'

const AccomCard = (props) => {

    const { onClick, accom } = props;
    const { id, name, starRating, img } = accom;

    return (
        <div className="accom_card" style={{ backgroundColor: 'gray', cursor: 'pointer' }} onClick={onClick}>
            
                <img src={img} alt={'accom_image' + id}/>
                
                <div className="details row">
                    <h3>{name}</h3>
                    <p>{starRating} stars</p>
                </div>
        </div>
    )   
}

export default AccomCard;