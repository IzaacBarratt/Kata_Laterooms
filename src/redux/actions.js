/*
 * action types
 */

export const ADD_SEARCH_FILTER      =   'ADD_SEARCH_FILTER'
export const REMOVE_SEARCH_FILTER   =   'REMOVE_SEARCH_FILTER'
export const CLEAR_SEARCH_FILTER    =   'CLEAR_SEARCH_FILTER'

export const ADD_ACCOM_ITEM         =   'ADD_ACCOM_ITEM'




export function addSearchFilter(filter) {
  return { type: ADD_SEARCH_FILTER, filter }
}

export function removeSearchFilter(filter) {
    return { type: REMOVE_SEARCH_FILTER, filter }
}

export function clearSearchFilter() {
    return { type: CLEAR_SEARCH_FILTER }
}
  
  


export function addAccomItem(data) {
  return { type: ADD_ACCOM_ITEM, data }
}

