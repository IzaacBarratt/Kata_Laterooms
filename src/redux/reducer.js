import {
  ADD_ACCOM_ITEM,
  ADD_SEARCH_FILTER,
  REMOVE_SEARCH_FILTER,
  CLEAR_SEARCH_FILTER
} from './actions'


const initialState = {
  filters: ['pool', 'gym'],
  accom: [
    {
      id: 1,
      "name": "hotelone",
      img: 'http://ihg.scene7.com/is/image/ihg/holiday-inn-the-colony-4629618286-4x3',
      "starRating": 5,
      "facilities": ["car park", "pool"]
    },
    {
      id: 2,
      "name": "hoteltwo",
      img: 'https://s-ec.bstatic.com/images/hotel/max1024x768/681/68184730.jpg',
      "starRating": 3,
      "facilities": ["car park", "gym"]
    },
    {
      id: 3,
      "name": "hotelthree",
      img: 'https://s-ec.bstatic.com/images/hotel/max1024x768/147/147997361.jpg',
      "starRating": 3,
      "facilities": ['bird farm']
    }
  ]
}



export default function accomApp(state = initialState, action) {

  console.log(action)
    
  switch (action.type) {

      case ADD_ACCOM_ITEM:
        const data = action.data;
        var nwState = Object.assign({}, state)

        if (Array.isArray(data)) {
          nwState.accom = nwState.accom.concat(data)
        } else {
          nwState.accom.push(data)
        }

        return nwState;
      

      // search filters ----------------------

      case ADD_SEARCH_FILTER:
        var filter = action.filter;
        var nwState = Object.assign({}, state)
        nwState.filters.push(filter)

        console.log(nwState)
        return nwState;

      case REMOVE_SEARCH_FILTER: 
        var filter = action.filter;
        var nwState = Object.assign({}, state);
        var filters = nwState.filters;
        
        // Check through filters and remove filter if matches
        for (var i = 0; i < filters.length; i ++) {
          if (filters[i] === filter) {
            filters.splice(i, 1);
            break;
          }
        }

        return nwState;

      case CLEAR_SEARCH_FILTER: 
        return Object.assign({}, state, { filters: [] })

      default:
        return state;
    }
  }