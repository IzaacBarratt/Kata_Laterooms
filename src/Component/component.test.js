import { createStore } from 'redux'
import React from 'react'
import reducer from '../redux/reducer'
import renderer from 'react-test-renderer'

import { 
    filterAccom
} from '../Utils/index'


const filter_item = 'helicopter pad'
const initial_state = {
    filters: [
        filter_item,
        'jacuuzi',
        'hospital'
    ],
    accom: [
        {
            id: 1,
            "name": "hotelone",
            "starRating": 5,
            "facilities": ["car park", "pool"]
        },
        {
            id: 2,
            "name": "hoteltwo",
            "starRating": 3,
            "facilities": ["car park", "gym"]
        },
        {
            id: 3,
            "name": "hotelthree",
            "starRating": 3,
            "facilities": []
        }
    ]
}

const store = createStore(reducer, initial_state);
import { 
    addSearchFilter,
    removeSearchFilter,
    clearSearchFilter
} from '../redux/actions'
import AccomCard from './AccomCard';


describe('Accom Items displaying', () => {

  it('matches using item 0', () => {
    const item = initial_state.accom[0];
    const tree = renderer.create(<AccomCard accom={item} />);
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when using item 1', () => {
    const item = initial_state.accom[1];
    const tree = renderer.create(<AccomCard accom={item}/>);
    expect(tree).toMatchSnapshot();
  });

});

test('Changing filter reflects in redux cache', () => {

    // Try adding to state
    const newFilter = 'pool'
    var state = store.getState();

   
    store.dispatch(addSearchFilter(newFilter))
    
    expect.objectContaining({'filters': []});
    expect(state.filters).toEqual(expect.arrayContaining([newFilter]));    
    
    
    // Try removing
    store.dispatch(removeSearchFilter(filter_item));

    var state = store.getState();

    expect.objectContaining({'filters': []});
    expect(state.filters).toEqual(expect.not.arrayContaining([filter_item]))

    // Try clearing 
    store.dispatch(clearSearchFilter());

    var state = store.getState();

    expect.objectContaining({'filters': []});
    expect(state.filters).toEqual([])

})


test('Filter function removes items that aren\'t queried', () => {

    const state = store.getState();
    var filters = state.filters;
    var accom = state.accom;

    var output = filterAccom(filters, accom)
    expect(output).toEqual([])

    output = filterAccom(filters.concat('car park'), accom)
    expect(output.length).toEqual(2)

    output = filterAccom(filters.concat('gym'), accom)
    expect(output.length).toEqual(1)

})



