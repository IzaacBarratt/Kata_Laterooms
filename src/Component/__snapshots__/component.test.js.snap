// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accom Items displaying matches snapshot when using item 1 1`] = `
<div
  className="accom_card"
  style={
    Object {
      "backgroundColor": "gray",
      "cursor": "pointer",
    }
  }
>
  <img
    alt="accom_image2"
  />
  <div
    className="details row"
  >
    <h3>
      hoteltwo
    </h3>
    <p>
      3
       stars
    </p>
  </div>
</div>
`;

exports[`Accom Items displaying matches using item 0 1`] = `
<div
  className="accom_card"
  style={
    Object {
      "backgroundColor": "gray",
      "cursor": "pointer",
    }
  }
>
  <img
    alt="accom_image1"
  />
  <div
    className="details row"
  >
    <h3>
      hotelone
    </h3>
    <p>
      5
       stars
    </p>
  </div>
</div>
`;

exports[`action creators creates CREATE_SUCCESS when creating a to-do was successful 1`] = `
Array [
  Object {
    "item": Object {
      "createdAt": "2017-03-02T23:04:38.003Z",
      "id": 1,
      "modifiedAt": "2017-03-22T16:44:29.034Z",
      "project": "Testing",
      "title": "Example",
    },
    "type": "CREATE_SUCCESS",
  },
]
`;
